### PR TITLE
Add transformer modeling docs

### PR DESCRIPTION
**Pull request by FrostML**
### PR types

Others

### PR changes

Docs

### Description

Add transformer modeling docs.

### CODE REVIEW

**Comment by FrostML** [comment]
发现 文档预览 出来有问题：
若未提供代码注释，默认会使用框架的 docstring 注释作为文档，以 `step` 为例。
这里的 `step`，严格来讲不需要用户显式调用，故未提供文档，但是预览出来会出现框架的文档，问题是：**和当前 PaddleNLP 文档有差别，文档的风格不能统一**。
image

**Review by ZeyuChen — APPROVED** [review]
LGTM

File: paddlenlp/transformers/transformer/modeling.py [hunk 1]
```diff
@@ -7,10 +7,27 @@
...
+    "position_encoding_init", "WordEmbedding", "PositionalEmbedding",
+    "CrossEntropyCriterion", "TransformerDecodeCell",
+    "TransformerBeamSearchDecoder", "TransformerModel", "InferTransformerModel"
+]
+
 
 def position_encoding_init(n_position, d_pos_vec, dtype="float32"):
     """
-    Generate the initial values for the sinusoid position encoding table.
+    Generate the initial values for the sinusoidal position encoding table.
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:19]: 统一第三人称，下同

File: paddlenlp/transformers/transformer/modeling.py [hunk 2]
```diff
@@ -7,10 +7,27 @@
...
+    different from the description in "Attention Is All You Need".
+
+    Args:
+        n_position (`int`): 
+            The largest position for sequences, that is, the maximum length
+            of source or target sequences.
+        d_pos_vec (`int`): 
+            The size of positional embedding vector. 
+        dtype (`str`): 
+            The output `numpy.array`'s data type.
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:30]: 数据类型统一不加标点，下同

File: paddlenlp/transformers/transformer/modeling.py [hunk 3]
```diff
@@ -7,10 +7,27 @@
...
+    Generate the initial values for the sinusoidal position encoding table.
+    This method follows the implementation in tensor2tensor, but is slightly
+    different from the description in "Attention Is All You Need".
+
+    Args:
+        n_position (`int`): 
+            The largest position for sequences, that is, the maximum length
+            of source or target sequences.
+        d_pos_vec (`int`): 
+            The size of positional embedding vector.
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:28]: The size of positional embedding vector. 
->
Dimensionality of positional embedding vector.

File: paddlenlp/transformers/transformer/modeling.py (hunk 2 above)
Comment (yingyibiao) [reply paddlenlp/transformers/transformer/modeling.py:30]: 缺少返回值，下同

File: paddlenlp/transformers/transformer/modeling.py [hunk 4]
```diff
@@ -44,16 +79,30 @@ def __init__(self, vocab_size, emb_dim, bos_idx=0):
                 initializer=nn.initializer.Normal(0., emb_dim**-0.5)))
 
     def forward(self, word):
+        r"""
+        Compute word embedding.
+
+        Args:
+            word (`Tensor`):
+                The input ids which means the sequences' words. 
+        """
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:88]: specify Tensor's shape and dtype

File: paddlenlp/transformers/transformer/modeling.py [hunk 5]
```diff
@@ -65,18 +114,46 @@ def __init__(self, emb_dim, max_length, bos_idx=0):
...
+    """
+    Compute the cross entropy loss for given input.
+
+    Args:
+        label_smooth_eps (`float`):
+            The weight used to mix up the original ground-truth distribution
+            and the fixed distribution. Default is None. If given, label smoothing
+            will be applied on `label`.
+        pad_idx (`int`):
+            The token id used to pad variant sequence. Default is 0.
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:139]: pad_idx缺少optional；另外，这里的描述有点问题，应该指出这里的pad_idx是算loss的时候ignore的token id。

File: paddlenlp/transformers/transformer/modeling.py [hunk 6]
```diff
@@ -99,6 +176,30 @@ def forward(self, predict, label):
 
 
 class TransformerDecodeCell(nn.Layer):
+    """
+    This layer wraps a Transformer decoder combined with embedding
+    layer and output layer to produce logits from ids and position.
+
+    Args:
+        decoder (`callable`):
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:184]: 这里的callable应该更加具体一点

File: paddlenlp/transformers/transformer/modeling.py [hunk 7]
```diff
@@ -99,6 +176,30 @@ def forward(self, predict, label):
...
+            includes a embedding layer. Default is None.
+        pos_embedding (`callable`, optional):
+            A `PositionalEmbedding` instance or a callable that accepts position
+            as arguments and return embeddings. It can be None if `decoder`
+            includes a positional embedding layer. Default is None.
+        linear (`callable`, optional):
+            A `paddle.nn.Linear` instance or a callable to transform decoder
+            output to logits.
+        dropout (`float`, optional):
+            The dropout rate for the results of `word_embedding` and `pos_embedding`.
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:200]: 缺少默认值

File: paddlenlp/transformers/transformer/modeling.py [hunk 8]
```diff
@@ -113,6 +214,31 @@ def __init__(self,
         self.dropout = dropout
 
     def forward(self, inputs, states, static_cache, trg_src_attn_bias, memory):
+        r"""
+        Produces logits.
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:218]: 介绍太简单

File: paddlenlp/transformers/transformer/modeling.py [hunk 9]
```diff
@@ -347,6 +577,13 @@ def __init__(self,
             cell, bos_id, eos_id, beam_size, var_dim_in_state=2)
 
     def forward(self, src_word):
+        r"""
+        The Transformer forward method.
+
+        Args:
+            src_word (`Tensor`):
+                The ids of source sequence words. 
+        """
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:586]: 缺少返回值

File: paddlenlp/transformers/transformer/modeling.py [hunk 10]
```diff
@@ -347,6 +577,13 @@ def __init__(self,
             cell, bos_id, eos_id, beam_size, var_dim_in_state=2)
 
     def forward(self, src_word):
+        r"""
+        The Transformer forward method.
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:581]: 这里介绍需要详细一点

File: paddlenlp/transformers/transformer/modeling.py [hunk 11]
```diff
@@ -347,6 +577,13 @@ def __init__(self,
...
+
+        Args:
+            src_word (`Tensor`):
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:584]: Tensor 参数需要介绍 shape and dtype

File: paddlenlp/transformers/transformer/modeling.py [hunk 12]
```diff
@@ -317,6 +513,40 @@ def forward(self, src_word, trg_word):
...
+            Whether to use weight sharing. 
+        bos_id (`int`, optional):
+            The start token id and also is used as padding id. Default is 0.
+        eos_id (`int`, optional):
+            The end token id. Default is 1.
+        beam_size (`int`, optional):
+            The beam width for beam search. Default is 4. 
+        max_out_len (`int`, optional):
+            The maximum output length. Default is 256.
+    """
```
Comment (yingyibiao) [inline paddlenlp/transformers/transformer/modeling.py:548]: 参数介绍可以参考xlnet.modeling的docstrings，需要详细一点；比如dropout是哪些结构的Dropout rates等。

**Comment by yingyibiao** [comment]
缺少代码示例

File: paddlenlp/transformers/transformer/modeling.py (hunk 5 above)
Comment (FrostML) [reply paddlenlp/transformers/transformer/modeling.py:139]: 不是，这里的 pad idx 就只是 padding index。

File: paddlenlp/transformers/transformer/modeling.py (hunk 6 above)
Comment (FrostML) [reply paddlenlp/transformers/transformer/modeling.py:184]: 这里理论上是各种 callable 都行，不固定，不需要是某一个类的 instance。太具体的话，就限制了这里的用法。

File: paddlenlp/transformers/transformer/modeling.py (hunk 10 above)
Comment (FrostML) [reply paddlenlp/transformers/transformer/modeling.py:581]: 这里是参考的 xlnet...
我再加点什么吧
